File: widgets/macos/Core/Sources/NankivWidgetModel/Snapshot.swift
import Foundation

// The snapshot the app publishes for its widgets.
//
// This mirrors `WidgetSnapshot` in src-tauri/src/widget.rs and is pinned by the
// files in widgets/fixtures, which the Rust tests generate and the tests here
// decode. Nothing in this file interprets data — every verdict, count and
// cutoff was decided by the app. The only judgement made on this side is how
// to survive a value this version does not recognise, and that judgement always
// errs toward "unknown", never toward an answer.

public struct WidgetSnapshot: Decodable, Sendable, Equatable {
    /// The schema this widget understands. A snapshot with any other schema was
    /// written by a different version of nankiv and is not trusted.
    public static let supportedSchema = 1

    public let schema: Int
    public let generatedAt: Date
    public let identityConfigured: Bool
    public let activity: Activity
    public let season: Season
    /// When the newest result stops leading the widget. `nil` when there is
    /// nothing to lead with.
    public let leadsUntil: Date?
    /// Newest first.
    public let drives: [DrivePreview]

    public init(
        schema: Int = WidgetSnapshot.supportedSchema,
        generatedAt: Date,
        identityConfigured: Bool,
        activity: Activity,
        season: Season,
        leadsUntil: Date? = nil,
        drives: [DrivePreview]
    ) {
        self.schema = schema
        self.generatedAt = generatedAt
        self.identityConfigured = identityConfigured
        self.activity = activity
        self.season = season
        self.leadsUntil = leadsUntil
        self.drives = drives
    }
}

public enum Activity: Decodable, Sendable, Equatable {
    case idle
    /// Shown until `until`, after which the app is assumed to have stopped.
    case importing(filename: String, since: Date, until: Date)
    /// Shown until `until`, or until an import succeeds.
    case failed(filename: String, at: Date, until: Date)

    private enum CodingKeys: String, CodingKey {
        case state, filename, since, until, at
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(String.self, forKey: .state) {
        case "importing":
            self = .importing(
                filename: try c.decode(String.self, forKey: .filename),
                since: try c.decode(Date.self, forKey: .since),
                until: try c.decode(Date.self, forKey: .until)
            )
        case "failed":
            self = .failed(
                filename: try c.decode(String.self, forKey: .filename),
                at: try c.decode(Date.self, forKey: .at),
                until: try c.decode(Date.self, forKey: .until)
            )
        default:
            // Something newer than this widget knows about. Showing the data
            // without a status line is better than showing nothing.
            self = .idle
        }
    }
}

public struct Season: Decodable, Sendable, Equatable {
    public let drives: Int
    public let shortlisted: Int
    public let notShortlisted: Int
    public let undetermined: Int

    public init(drives: Int, shortlisted: Int, notShortlisted: Int, undetermined: Int) {
        self.drives = drives
        self.shortlisted = shortlisted
        self.notShortlisted = notShortlisted
        self.undetermined = undetermined
    }
}

public struct DrivePreview: Decodable, Sendable, Equatable, Identifiable {
    public let id: Int64
    public let company: String
    public let importedAt: Date
    public let totalStudents: Int
    public let verdict: Verdict
    public let circle: CircleSummary
    public let analysis: AnalysisSummary

    public init(
        id: Int64,
        company: String,
        importedAt: Date,
        totalStudents: Int,
        verdict: Verdict,
        circle: CircleSummary,
        analysis: AnalysisSummary
    ) {
        self.id = id
        self.company = company
        self.importedAt = importedAt
        self.totalStudents = totalStudents
        self.verdict = verdict
        self.circle = circle
        self.analysis = analysis
    }
}

/// Three answers, never two. There is deliberately no `Bool` view of this.
public enum Verdict: Decodable, Sendable, Equatable {
    case shortlisted
    case notShortlisted
    case undetermined(UndeterminedReason)

    private enum CodingKeys: String, CodingKey { case status, reason }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let reason = try c.decodeIfPresent(String.self, forKey: .reason)
        self.init(status: try c.decode(String.self, forKey: .status), reason: reason)
    }

    init(status: String, reason: String?) {
        switch status {
        case "shortlisted": self = .shortlisted
        case "not_shortlisted": self = .notShortlisted
        default:
            // Includes any status this version does not recognise: an unknown
            // word is never read as a rejection.
            self = .undetermined(reason.flatMap(UndeterminedReason.init(rawValue:)) ?? .other)
        }
    }
}

public enum UndeterminedReason: String, Sendable, Equatable {
    case noIdentityConfigured = "no_identity_configured"
    case needsNeoId = "needs_neo_id"
    case needsRegNo = "needs_reg_no"
    case fileNotUnderstood = "file_not_understood"
    case other
}

public struct CircleSummary: Decodable, Sendable, Equatable {
    public let total: Int
    public let shortlisted: Int
    public let notShortlisted: Int
    public let undetermined: Int
    /// In the app's order: shortlisted, then undetermined, then not shortlisted.
    public let members: [Member]

    public init(total: Int, shortlisted: Int, notShortlisted: Int, undetermined: Int, members: [Member]) {
        self.total = total
        self.shortlisted = shortlisted
        self.notShortlisted = notShortlisted
        self.undetermined = undetermined
        self.members = members
    }

    public static let none = CircleSummary(total: 0, shortlisted: 0, notShortlisted: 0, undetermined: 0, members: [])

    public struct Member: Decodable, Sendable, Equatable {
        public let label: String
        public let verdict: Verdict

        public init(label: String, verdict: Verdict) {
            self.label = label
            self.verdict = verdict
        }

        private enum CodingKeys: String, CodingKey { case label, status }

        public init(from decoder: any Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            label = try c.decode(String.self, forKey: .label)
            verdict = Verdict(status: try c.decode(String.self, forKey: .status), reason: nil)
        }
    }
}

public struct AnalysisSummary: Decodable, Sendable, Equatable {
    public enum Status: String, Sendable, Equatable {
        case ready
        case insufficient
        case notSetUp = "not_set_up"
    }

    public let status: Status
    public let matched: Int
    public let coverage: Double
    /// Present only when `status` is `ready`.
    public let cutoff: Cutoff?
    public let median: Double?
    /// Empty buckets at either end already trimmed.
    public let histogram: [Bucket]
    public let overRepresented: [String]

    public init(
        status: Status,
        matched: Int,
        coverage: Double,
        cutoff: Cutoff?,
        median: Double?,
        histogram: [Bucket],
        overRepresented: [String]
    ) {
        self.status = status
        self.matched = matched
        self.coverage = coverage
        self.cutoff = status == .ready ? cutoff : nil
        self.median = status == .ready ? median : nil
        self.histogram = status == .ready ? histogram : []
        self.overRepresented = status == .ready ? overRepresented : []
    }

    private enum CodingKeys: String, CodingKey {
        case status, matched, coverage, cutoff, median, histogram, overRepresented
    }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let raw = try c.decode(String.self, forKey: .status)
        self.init(
            // An unfamiliar status shows no statistics rather than guessed ones.
            status: Status(rawValue: raw) ?? .insufficient,
            matched: try c.decode(Int.self, forKey: .matched),
            coverage: try c.decode(Double.self, forKey: .coverage),
            // A cutoff this version cannot read is left out, not allowed to
            // fail the whole snapshot.
            cutoff: (try? c.decodeIfPresent(Cutoff.self, forKey: .cutoff)) ?? nil,
            median: try c.decodeIfPresent(Double.self, forKey: .median),
            histogram: try c.decodeIfPresent([Bucket].self, forKey: .histogram) ?? [],
            overRepresented: try c.decodeIfPresent([String].self, forKey: .overRepresented) ?? []
        )
    }
}

/// Every case is an estimate. The widget labels it as one wherever it appears,
/// exactly as the app does: a pattern in who was shortlisted is not a rule the
/// company published.
public enum Cutoff: Sendable, Equatable {
    case around(threshold: Double, observedFloor: Double)
    case skewsHigh(observedFloor: Double)
    case noFilter
}

extension Cutoff: Decodable {
    private enum CodingKeys: String, CodingKey { case kind, threshold, observedFloor }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let threshold = try c.decodeIfPresent(Double.self, forKey: .threshold)
        let floor = try c.decodeIfPresent(Double.self, forKey: .observedFloor)
        switch (try c.decode(String.self, forKey: .kind), threshold, floor) {
        case ("hard_cutoff", let t?, let f?): self = .around(threshold: t, observedFloor: f)
        case ("soft_preference", _, let f?): self = .skewsHigh(observedFloor: f)
        case ("no_cgpa_filter", _, _): self = .noFilter
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .kind, in: c, debugDescription: "unrecognised cutoff")
        }
    }
}

public struct Bucket: Decodable, Sendable, Equatable {
    public let lower: Double
    public let upper: Double
    public let count: Int

    public init(lower: Double, upper: Double, count: Int) {
        self.lower = lower
        self.upper = upper
        self.count = count
    }
}

File: widgets/windows/provider/Snapshot.cs
// The snapshot the app publishes, as C# sees it.
//
// A mirror of src-tauri/src/widget.rs, like Swift's NankivWidgetModel. The
// files in widgets/fixtures are the shared examples; Nankiv.Widgets.Tests
// decodes every one of them.
//
// Nothing here interprets data. Where a value is unrecognised the choice is
// always "unknown", never an answer — a shortlist result that cannot be read
// must not turn into "not shortlisted".

using System.Text.Json;
using System.Text.Json.Serialization;

namespace Nankiv.Widgets;

public sealed record WidgetSnapshot(
    int Schema,
    DateTimeOffset GeneratedAt,
    bool IdentityConfigured,
    Activity Activity,
    Season Season,
    DateTimeOffset? LeadsUntil,
    IReadOnlyList<DrivePreview> Drives)
{
    public const int SupportedSchema = 1;
}

public abstract record Activity
{
    public sealed record Idle : Activity;

    /// <summary>Shown until <paramref name="Until"/>: past it, the app is assumed to have stopped.</summary>
    public sealed record Importing(string Filename, DateTimeOffset Since, DateTimeOffset Until) : Activity;

    /// <summary>Shown until <paramref name="Until"/>, or until an import succeeds.</summary>
    public sealed record Failed(string Filename, DateTimeOffset At, DateTimeOffset Until) : Activity;
}

public sealed record Season(int Drives, int Shortlisted, int NotShortlisted, int Undetermined);

public sealed record DrivePreview(
    long Id,
    string Company,
    DateTimeOffset ImportedAt,
    int TotalStudents,
    Verdict Verdict,
    CircleSummary Circle,
    AnalysisSummary Analysis);

public enum VerdictStatus { Shortlisted, NotShortlisted, Undetermined }

public enum UndeterminedReason { NoIdentityConfigured, NeedsNeoId, NeedsRegNo, FileNotUnderstood, Other }

public sealed record Verdict(VerdictStatus Status, UndeterminedReason? Reason)
{
    public static Verdict From(string? status, string? reason) => status switch
    {
        "shortlisted" => new Verdict(VerdictStatus.Shortlisted, null),
        "not_shortlisted" => new Verdict(VerdictStatus.NotShortlisted, null),
        // Includes anything this version does not recognise.
        _ => new Verdict(VerdictStatus.Undetermined, reason switch
        {
            "no_identity_configured" => UndeterminedReason.NoIdentityConfigured,
            "needs_neo_id" => UndeterminedReason.NeedsNeoId,
            "needs_reg_no" => UndeterminedReason.NeedsRegNo,
            "file_not_understood" => UndeterminedReason.FileNotUnderstood,
            _ => UndeterminedReason.Other,
        }),
    };
}

public sealed record CircleSummary(
    int Total,
    int Shortlisted,
    int NotShortlisted,
    int Undetermined,
    IReadOnlyList<CircleMember> Members);

public sealed record CircleMember(string Label, Verdict Verdict);

public enum AnalysisStatus { Ready, Insufficient, NotSetUp }

public sealed record AnalysisSummary(
    AnalysisStatus Status,
    int Matched,
    double Coverage,
    Cutoff? Cutoff,
    double? Median,
    IReadOnlyList<Bucket> Histogram,
    IReadOnlyList<string> OverRepresented);

public enum CutoffKind { Around, SkewsHigh, NoFilter }

/// <summary>Always an estimate, and always presented as one.</summary>
public sealed record Cutoff(CutoffKind Kind, double? Threshold, double? ObservedFloor);

public sealed record Bucket(double Lower, double Upper, int Count);

public enum SnapshotStateKind { Missing, Stale, Loaded }

public sealed record SnapshotLoad(SnapshotStateKind Kind, WidgetSnapshot? Snapshot)
{
    public static readonly SnapshotLoad Missing = new(SnapshotStateKind.Missing, null);
    public static readonly SnapshotLoad Stale = new(SnapshotStateKind.Stale, null);
    public static SnapshotLoad Loaded(WidgetSnapshot s) => new(SnapshotStateKind.Loaded, s);
}

/// <summary>
/// Reads the snapshot the app writes. One small file; no database, no analysis.
/// </summary>
public static class SnapshotReader
{
    /// <summary>Must match the directory <c>lib.rs</c> publishes into.</summary>
    public static string DefaultPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "app.nankiv.desktop", "widget", "snapshot.json");

    public static SnapshotLoad Load(string? path = null)
    {
        path ??= DefaultPath;
        if (!File.Exists(path)) return SnapshotLoad.Missing;
        try
        {
            return Decode(File.ReadAllText(path));
        }
        catch (IOException)
        {
            // Being rewritten as we read: the writer renames into place, so the
            // next update will bring a whole file.
            return SnapshotLoad.Stale;
        }
    }

    public static SnapshotLoad Decode(string json)
    {
        try
        {
            using var document = JsonDocument.Parse(json);
            if (!document.RootElement.TryGetProperty("schema", out var schema)
                || schema.GetInt32() != WidgetSnapshot.SupportedSchema)
            {
                // Written by another version of nankiv. Opening the app rewrites it.
                return SnapshotLoad.Stale;
            }
            var snapshot = JsonSerializer.Deserialize<WidgetSnapshot>(json, Options);
            return snapshot is null ? SnapshotLoad.Stale : SnapshotLoad.Loaded(snapshot);
        }
        catch (JsonException)
        {
            return SnapshotLoad.Stale;
        }
    }

    public static readonly JsonSerializerOptions Options = Build();

    private static JsonSerializerOptions Build()
    {
        var options = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            PropertyNameCaseInsensitive = true,
        };
        options.Converters.Add(new ActivityConverter());
        options.Converters.Add(new VerdictConverter());
        options.Converters.Add(new CircleMemberConverter());
        options.Converters.Add(new AnalysisConverter());
        return options;
    }
}

internal sealed class ActivityConverter : JsonConverter<Activity>
{
    public override Activity Read(ref Utf8JsonReader reader, Type type, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        var root = document.RootElement;
        var state = root.TryGetProperty("state", out var s) ? s.GetString() : null;
        switch (state)
        {
            case "importing":
                return new Activity.Importing(
                    root.GetProperty("filename").GetString() ?? "",
                    root.GetProperty("since").GetDateTimeOffset(),
                    root.GetProperty("until").GetDateTimeOffset());
            case "failed":
                return new Activity.Failed(
                    root.GetProperty("filename").GetString() ?? "",
                    root.GetProperty("at").GetDateTimeOffset(),
                    root.GetProperty("until").GetDateTimeOffset());
            default:
                // Something newer than this widget knows about: show the data
                // without a status line rather than nothing at all.
                return new Activity.Idle();
        }
    }

    public override void Write(Utf8JsonWriter writer, Activity value, JsonSerializerOptions options) =>
        throw new NotSupportedException("the widget never writes snapshots");
}

internal sealed class VerdictConverter : JsonConverter<Verdict>
{
    public override Verdict Read(ref Utf8JsonReader reader, Type type, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        var root = document.RootElement;
        return Verdict.From(
            root.TryGetProperty("status", out var s) ? s.GetString() : null,
            root.TryGetProperty("reason", out var r) ? r.GetString() : null);
    }

    public override void Write(Utf8JsonWriter writer, Verdict value, JsonSerializerOptions options) =>
        throw new NotSupportedException("the widget never writes snapshots");
}

/// <summary>
/// A circle member carries its answer as a plain `status`, not as the nested
/// verdict object a drive has — there is no reason for it, since a friend's
/// result has no explanation attached. Without this the property decodes as
/// null and the first card built from it throws.
/// </summary>
internal sealed class CircleMemberConverter : JsonConverter<CircleMember>
{
    public override CircleMember Read(ref Utf8JsonReader reader, Type type, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        var root = document.RootElement;
        return new CircleMember(
            root.TryGetProperty("label", out var label) ? label.GetString() ?? "" : "",
            Verdict.From(root.TryGetProperty("status", out var status) ? status.GetString() : null, null));
    }

    public override void Write(Utf8JsonWriter writer, CircleMember value, JsonSerializerOptions options) =>
        throw new NotSupportedException("the widget never writes snapshots");
}

internal sealed class AnalysisConverter : JsonConverter<AnalysisSummary>
{
    public override AnalysisSummary Read(ref Utf8JsonReader reader, Type type, JsonSerializerOptions options)
    {
        using var document = JsonDocument.ParseValue(ref reader);
        var root = document.RootElement;
        var status = root.TryGetProperty("status", out var s) ? s.GetString() : null;
        var parsed = status switch
        {
            "ready" => AnalysisStatus.Ready,
            "not_set_up" => AnalysisStatus.NotSetUp,
            // An unfamiliar status shows no statistics rather than guessed ones.
            _ => AnalysisStatus.Insufficient,
        };

        Cutoff? cutoff = null;
        if (parsed == AnalysisStatus.Ready
            && root.TryGetProperty("cutoff", out var c)
            && c.ValueKind == JsonValueKind.Object)
        {
            var kind = c.TryGetProperty("kind", out var k) ? k.GetString() : null;
            double? threshold = c.TryGetProperty("threshold", out var t) && t.ValueKind == JsonValueKind.Number
                ? t.GetDouble() : null;
            double? floor = c.TryGetProperty("observed_floor", out var f) && f.ValueKind == JsonValueKind.Number
                ? f.GetDouble() : null;
            cutoff = (kind, threshold) switch
            {
                ("hard_cutoff", not null) => new Cutoff(CutoffKind.Around, threshold, floor),
                ("soft_preference", _) => new Cutoff(CutoffKind.SkewsHigh, null, floor),
                ("no_cgpa_filter", _) => new Cutoff(CutoffKind.NoFilter, null, null),
                // A cutoff this version cannot read is left out, not guessed at.
                _ => null,
            };
        }

        var histogram = new List<Bucket>();
        if (parsed == AnalysisStatus.Ready
            && root.TryGetProperty("histogram", out var h)
            && h.ValueKind == JsonValueKind.Array)
        {
            foreach (var bucket in h.EnumerateArray())
            {
                histogram.Add(new Bucket(
                    bucket.GetProperty("lower").GetDouble(),
                    bucket.GetProperty("upper").GetDouble(),
                    bucket.GetProperty("count").GetInt32()));
            }
        }

        var branches = new List<string>();
        if (parsed == AnalysisStatus.Ready
            && root.TryGetProperty("over_represented", out var o)
            && o.ValueKind == JsonValueKind.Array)
        {
            foreach (var branch in o.EnumerateArray())
            {
                if (branch.GetString() is { } name) branches.Add(name);
            }
        }

        return new AnalysisSummary(
            parsed,
            root.TryGetProperty("matched", out var m) ? m.GetInt32() : 0,
            root.TryGetProperty("coverage", out var cov) ? cov.GetDouble() : 0,
            cutoff,
            parsed == AnalysisStatus.Ready && root.TryGetProperty("median", out var med)
                && med.ValueKind == JsonValueKind.Number ? med.GetDouble() : null,
            histogram,
            branches);
    }

    public override void Write(Utf8JsonWriter writer, AnalysisSummary value, JsonSerializerOptions options) =>
        throw new NotSupportedException("the widget never writes snapshots");
}
